### PR TITLE
Remove dependency on Laminas\Crypt

### PR DESCRIPTION
**Pull request by gsteel**
This patch ships a copy of "Apache Password" from laminas-crypt with several simplifications and improved tests. It effectively removes all dependencies on laminas-crypt with the exception that it would be a BC break to remove the protected method `getApachePassword()` from the ApacheResolver.

Perhaps, with the given changes, it might be possible to remove the dep from composer, therefore requiring that anyone who's extended from ApacheResolver also install laminas-crypt as it is now completely 'un-used' here.

### CODE REVIEW

**Comment by Ocramius** [comment]
@ezimuel do you have time to check this, perhaps?

File: src/Adapter/Http/ApachePassword.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+     * Verify if a clear-text password is correct given a hash value
+     *
+     * The username and realm are required in order to check credentials in digest format.
+     */
+    public static function verify(string $password, string $hash, string|null $username, string|null $realm): bool
+    {
+        if (str_starts_with($hash, '{SHA}')) {
+            $hash2 = '{SHA}' . base64_encode(sha1($password, true));
+
+            return hash_equals($hash, $hash2);
```
Comment (lcobucci) [inline src/Adapter/Http/ApachePassword.php]: I might be quite wrong but I feel like adding prefixes and b64 encode remove all the timing attack protection offered by `hash_equals()`.

I'd suggest doing the inverse and removing the prefix + b64 decode on `$hash`:

```suggestion
            return hash_equals(
                base64_decode(substr($hash, 5), true),
                sha1($password, true)
            );
```

File: src/Adapter/Http/ApachePassword.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+declare(strict_types=1);
+
+namespace Laminas\Authentication\Adapter\Http;
+
+use function base64_encode;
+use function chr;
+use function count;
+use function explode;
+use function hash_equals;
+use function mb_strlen;
```
Comment (lcobucci) [inline src/Adapter/Http/ApachePassword.php:13]: Should we add `ext-mbstring` to composer? It seems like it's the first place we're adding `mb_*` functions.

**Comment by gsteel** [comment]
Thanks for the review @lcobucci - I've also dropped `laminas-crypt` from require-dev and suggest.

Do we need more reviews from @laminas/technical-steering-committee on this as it's security sensitive?

File: composer.json [hunk 3]
```diff
@@ -21,29 +21,28 @@
         },
         "sort-packages": true,
         "platform": {
-            "php": "8.1.99"
+            "php": "8.0.99"
         }
     },
     "require": {
-        "php": "~8.1.0 || ~8.2.0 || ~8.3.0",
+        "php": "~8.0.0 || ~8.1.0 || ~8.2.0",
```
Comment (froschdesign) [inline composer.json]: @gsteel 
Why?

Comment (gsteel) [reply composer.json]: My mistake - Just did a rebase and killed it. Hang on

Comment (froschdesign) [reply composer.json]: No problem. Good work! 👍🏻

**Comment by gsteel** [comment]
@weierophinney Could you possibly review this one for me when you get a minute? I'd like to get it off my desk and it needs final approval 🙏

**Review by lcobucci — APPROVED** [review]

**Pull request merged**